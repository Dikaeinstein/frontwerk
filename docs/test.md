# frontwerk 🛠️ tests your javascript

Test your JS files using [Jest][jest].

## Default configuration

By default, `frontwerk` will use the following jest rules:

```json
{
  "roots": "src",
  "testEnvironment":
    "jsdom (if not using webpack or rollup or react) or node otherwise",
  "collectCoverageFrom": ["src/**/*.js"],
  "testMatch": ["**/__tests__/**/*.js"],
  "testPathIgnorePatterns": [
    "/node_modules/",
    "/fixtures/",
    "/__tests__/helpers/",
    "__mocks__"
  ],
  "coveragePathIgnorePatterns": [
    "/node_modules/",
    "/fixtures/",
    "/__tests__/helpers/",
    "__mocks__",
    "src/(umd|cjs|esm)-entry.js$"
  ],
  "transformIgnorePatterns": ["[/\\\\]node_modules[/\\\\].+\\.(js|jsx)$"],
  "coverageThreshold": {
    "global": {
      "branches": 80,
      "functions": 80,
      "lines": 80,
      "statements": 80
    }
  }
}
```

## Overriding the defaults

`frontwerk` allows for easy extension and configuration of this default config.

> Note that any args you pass to `frontwerk test` will be forwarded to jest.

**Config**

There are two possible ways to extend or create your jest config.

1. Create a file named `jest.config.js` in your project root and pass it using
   the `--config` to your test task.
2. Have a `jest` property in your `package.json`.

You can override the default config by creating a `jest.config.js` file.

```javascript
const { jest: jestConfig } = require('frontwerk/config');
module.exports = Object.assign(jestConfig, {
  // your overrides here
});
```

Add pass a `--config` flag to your `frontwerk format` task with a path to a file
to use as a configuration file.

```json
{
  "scripts": {
    "test": "frontwerk test --config ./jest.config.js"
  }
}
```

**Options**

There are a few CLI flags you can use:

* `--no-watch` will not watch the files for changes. The default is `watch`.
* `--coverage` will display a coverage report and also override `watch`.
* `--updateSnapshot` will update snapshots and also override `watch`.

[jest]: https://facebook.github.io/jest/
