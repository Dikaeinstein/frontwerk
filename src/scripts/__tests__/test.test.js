/* eslint-disable global-require */
jest.mock('jest', () => ({ run: jest.fn() }));
jest.mock('../../config/jest.config', () => ({ builtInConfig: true }));

const cases = require('jest-in-case');

const testCases = [
  {
    name: 'calls jest CLI with default args'
  },
  {
    name: 'does not watch with --no-watch',
    args: ['--no-watch']
  },
  {
    name: 'does not watch with --coverage',
    args: ['--no-coverage']
  },
  {
    name: 'does not watch --updateSnapshot',
    args: ['--updateSnapshot']
  },
  {
    name: 'does not use built-in config with jest pkg prop',
    hasPkgProp: prop => prop === 'jest'
  },
  {
    name: 'does not use built-in config with --config',
    args: ['--config', './custom-config.js']
  },
  {
    name: 'forwards args',
    args: ['--coverage', '--watch']
  }
];

const testFn = ({ hasPkgProp = () => false, args = [] }) => {
  const { run: jestRunMock } = require('jest');
  const originalExit = process.exit;
  const originalArgv = process.argv;
  process.exit = jest.fn();

  Object.assign(require('../../utils/pkg'), { hasPkgProp });

  process.exit = jest.fn();

  try {
    // tests
    process.argv = ['node', '../test', ...args];
    jestRunMock.mockClear();

    require('../test');

    expect(jestRunMock).toHaveBeenCalledTimes(1);
    const [firstCall] = jestRunMock.mock.calls;
    const [jestArgs] = firstCall;
    expect(jestArgs.join(' ')).toMatchSnapshot();
  } catch (error) {
    throw error;
  } finally {
    // afterEach
    process.exit = originalExit;
    process.argv = originalArgv;
    jest.resetModules();
  }
};

cases('format', testFn, testCases);
