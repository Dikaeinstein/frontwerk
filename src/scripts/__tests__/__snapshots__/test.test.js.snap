// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`format calls jest CLI with default args 1`] = `"--config {\\"builtInConfig\\":true} --watch"`;

exports[`format does not use built-in config with --config 1`] = `"--watch --config ./custom-config.js"`;

exports[`format does not use built-in config with jest pkg prop 1`] = `"--watch"`;

exports[`format does not watch --updateSnapshot 1`] = `"--config {\\"builtInConfig\\":true} --updateSnapshot"`;

exports[`format does not watch with --coverage 1`] = `"--config {\\"builtInConfig\\":true} --watch --no-coverage"`;

exports[`format does not watch with --no-watch 1`] = `"--config {\\"builtInConfig\\":true} --no-watch"`;

exports[`format forwards args 1`] = `"--config {\\"builtInConfig\\":true} --coverage --watch"`;
